controllers/s3utils.go

```
@@ -0,0 +1,404 @@
+func (s *s3Connection) createBucket(bucket string) error {
+	if bucket == "" {
+		return fmt.Errorf("error: unspecified bucket name for "+
+			"endpoint %v tag %v",
+			s.endpoint, s.callerTag)
+	}
+
+	_, err := s.client.CreateBucket(
+		&s3.CreateBucketInput{Bucket: &bucket})
+	if err != nil {
+		if !errorswrapper.As(err, s3.ErrCodeBucketAlreadyOwnedByYou) {
+			return errorswrapper.Wrapf(err, "Failed to create bucket %s",
+				bucket)
+		}
+	}
+
+	return nil
+}
+
+// A convenience method
+// - OK to call UploadPV() concurrently from multiple goroutines safely.
+func (s *s3Connection) uploadPV(bucket string, pvKeySuffix string,
+	pv corev1.PersistentVolume) error {
+	return s.uploadTypedObject(bucket, pvKeySuffix /* key suffix */, pv)
+}
+
+// Upload to the given bucket the given uploadContent with a key of
+// <objectType/keySuffix>, where objectType is the type of the uploadContent
+// parameter
+// - OK to call UploadTypedObject() concurrently from multiple goroutines safely.
+func (s *s3Connection) uploadTypedObject(bucket string, keySuffix string,
+	uploadContent interface{}) error {
+	keyPrefix := reflect.TypeOf(uploadContent).String() + "/"
+	key := keyPrefix + keySuffix
+
+	return s.uploadObject(bucket, key, uploadContent)
+}
+
+// Upload the given object to the given bucket with the given key
+// - OK to call UploadObject() concurrently from multiple goroutines safely.
+// - Upload may fail due to many reasons: RequestError (connection error),
+//	 NoSuchBucket, NoSuchKey, InvalidParameter (e.g., empty key), etc.
+// - Multiple consecutive forward slashes in the key are sqaushed to
+//	 a single forward slash, for each such occurrence
+// - Any formatting changes to this method should also be reflected in the
+//	 downloadObject() method
+func (s *s3Connection) uploadObject(bucket string, key string,
+	uploadContent interface{}) error {
+	encodedUploadContent := &bytes.Buffer{}
+
+	gzWriter := gzip.NewWriter(encodedUploadContent)
```

Probably add `defer gzWriter.Close()` instead of having it in line 233.

---

Given that I want to return an err if close fails,  I need to defer a func() to change the return value.  Is that what you mean?